src/dmd/func.d

```diff
@@ -3749,7 +3749,18 @@ extern (C++) final class FuncLiteralDeclaration : FuncDeclaration
                 Expression exp = s.exp;
                 if (exp && !exp.type.equals(tret))
                 {
-                    s.exp = exp.castTo(sc, tret);
+                    /* https://issues.dlang.org/show_bug.cgi?id=23234
+
+                    If the return statement of a FuncLiteralDeclaration
+                    is rewritten to match an expected type, we should
+                    check alias this before bluntly casting.
+                    */
+                    import dmd.aliasthis : resolveAliasThis;
```

Comment:
The change in behaviour needs to be documented at the top of modifyReturns too

Reply:
Also can anything fail at this point semantically?

Reply:
(To be clear, this is not so much a change in behavior as a miscompilation. The existing code silently creates crashes.)

Reply:
I think that the idea here was that errors should be avoided. `implicitCastTo` may issue some errors if that is not possible, whereas with the current implementation, the return expression is simply casted to the expected type. If it were me, I would just drop the callsite-based return type reference.

Reply:
I mean, but shouldn't the return type inference already be based on what sort of types can be implicitly converted?  For instance-
```
void foo(bool delegate() dg) { }
void main() { foo({ return 5; }); }
```
This errors, even though 5 can be `castTo` bool.